### ISSUE TITLE
Type of LHS of walrus operator not correctly inferred

### ISSUE THREAD

**Issue by turlando** [python/mypy#14407]
**Bug Report**

The type of the LHS of an assignment performed using the walrus operator is not correctly inferred.
This issue can lead to insidious bugs where the user forgets to add a pair of brackets around the assignment.

**To Reproduce**

```python
FOO: dict[str, int] = {
    'a': 1,
    'b': 2
}


def foo(a: str) -> int:
    if x := FOO.get(a) is not None:
        return x
    return -1


if __name__ == '__main__':
    print(foo('a'))
```

```
% python ~/testcase.py
True
```

**Expected Behavior**

mypy should report that it is not possible to unify int (coming from the return type of `foo`) and bool (coming from the `is` operator).

This can lead to bugs where the user does not realize that they should have really written `if (x := FOO.get(a)) is not None`.

**Actual Behavior**

```
% mypy ~/testcase.py  
Success: no issues found in 1 source file
```

**Your Environment**

- Mypy version used: `mypy 0.991 (compiled: yes)`
- Mypy command-line flags: _none_
- Mypy configuration options from `mypy.ini` (and other config files): _none_
- Python version used: Python 3.10.8

**Comment by hauntsaninja** [comment]
I'm sympathetic, but bool is a subtype of int
